File: implementations/python/sampling/reservoir/reservoir.py
import random

def reservoir_sample(stream, k):
    if k <= 0:
        return []
    if len(stream) <= k:
        return list(stream)
    reservoir = list(stream[:k])
    for i in range(k, len(stream)):
        j = random.randint(0, i)
        if j < k:
            reservoir[j] = stream[i]
    return reservoir

File: implementations/javascript/sampling/reservoir/reservoir.js
export function reservoirSample(stream, k) {
  if (k <= 0) return [];
  if (stream.length <= k) return [...stream];
  const reservoir = stream.slice(0, k);
  for (let i = k; i < stream.length; i++) {
    const j = Math.floor(Math.random() * (i + 1));
    if (j < k) {
      reservoir[j] = stream[i];
    }
  }
  return reservoir;
}
